File: neet_code_150_cpp/stack/rev_polish.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>

class Solution {
    public:
        int eval_rpn(std::vector<std::string>& toknes) {
            std::stack<int> stack;

            for(const auto &c: toknes) {
                if(c == "+") {
                    int a = stack.top();
                    stack.pop();
                    int b = stack.top();
                    stack.pop();
                    stack.push(b + a);
                }else if(c == "-") {
                    int a = stack.top();
                    stack.pop();
                    int b = stack.top();
                    stack.pop();
                    stack.push(b - a);
                }else if(c == "*") {
                    int a = stack.top();
                    stack.pop();
                    int b = stack.top();
                    stack.pop();
                    stack.push(b * a);
                }else if(c == "/") {
                    int a = stack.top();
                    stack.pop();
                    int b = stack.top();
                    stack.pop();
                    stack.push(b / a);
                } else {
                    stack.push(std::stoi(c));
                }
            }

            std::cout << stack.top() << std::endl;
            return stack.top();
        }
};

class SolutionAns {
public:
    int evalRPN(std::vector<std::string>& tokens) {
        std::stack<int> stk;
        
        for (int i = 0; i < static_cast<int>(tokens.size()); i++) {
            std::string token = tokens[i];
            
            if (token.size() > 1 || isdigit(token[0])) {
                stk.push(stoi(token));
                continue;
            }
            
            int num2 = stk.top();
            stk.pop();
            int num1 = stk.top();
            stk.pop();
            
            int result = 0;
            if (token == "+") {
                result = num1 + num2;
            } else if (token == "-") {
                result = num1 - num2;
            } else if (token == "*") {
                result = num1 * num2;
            } else {
                result = num1 / num2;
            }
            stk.push(result);
        }
        
        std::cout << stk.top() << std::endl;
        return stk.top();
    }
};

int main(void) {
    std::vector<std::string> case_1 = {"4", "13", "5", "/", "+"};
    std::vector<std::string> case_2 = {"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"};

    Solution s_1;
    s_1.eval_rpn(case_1);
    s_1.eval_rpn(case_2);

    SolutionAns s_ans;
    s_ans.evalRPN(case_1);
    s_ans.evalRPN(case_2);
}

File: neet_code_150/src/stack/rev_polish.rs
struct Solution {}

// AC
impl Solution {
    pub fn eval_rpn(tokens: Vec<String>) -> i32 {
        let mut numbers = vec![];
        for c in tokens {
            match c.parse::<i32>() {
                Ok(num) => numbers.push(num),
                Err(_e) => {
                    let a = match numbers.pop() {
                        Some(num) => num,
                        None => {
                            panic!("error");
                        }
                    };
                    let b = match numbers.pop() {
                        Some(num) => num,
                        None => {
                            panic!("error");
                        }
                    };
                    match c.as_str() {
                        "+" => {
                            let result = b + a;
                            numbers.push(result);
                        }
                        "-" => {
                            let result = b - a;
                            numbers.push(result);
                        }
                        "*" => {
                            let result = b * a;
                            numbers.push(result);
                        }
                        "/" => {
                            let result = b / a;
                            numbers.push(result);
                        }
                        _ => {
                            panic!("invalid operator");
                        }
                    }
                }
            }
        }
        println!("{}", numbers.iter().rev().next().unwrap());
        match numbers.iter().rev().next() {
            Some(num) => *num,
            None => {
                panic!("error");
            }
        }
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn eval_rpn(tokens: Vec<String>) -> i32 {
        let mut stack: Vec<i32> = Vec::new();

        for token in tokens {
            match &token[..] {
                "+" => {
                    let second_operand = stack.pop().unwrap();
                    let first_operand = stack.pop().unwrap();
                    stack.push(first_operand + second_operand)
                }
                "-" => {
                    let second_operand = stack.pop().unwrap();
                    let first_operand = stack.pop().unwrap();
                    stack.push(first_operand - second_operand)
                }
                "*" => {
                    let second_operand = stack.pop().unwrap();
                    let first_operand = stack.pop().unwrap();
                    stack.push(first_operand * second_operand)
                }
                "/" => {
                    let second_operand = stack.pop().unwrap();
                    let first_operand = stack.pop().unwrap();
                    stack.push(first_operand / second_operand)
                }
                value => stack.push(value.parse::<i32>().unwrap()),
            }
        }

        println!("{}", stack[0]);
        stack[0]
    }
}

fn main() {
    let case_1: Vec<String> = vec!["4", "13", "5", "/", "+"]
        .iter()
        .map(|s| s.to_string())
        .collect();

    let case_2: Vec<String> = vec![
        "10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+",
    ]
    .iter()
    .map(|s| s.to_string())
    .collect();

    Solution::eval_rpn(case_1.clone());
    Solution::eval_rpn(case_2.clone());

    SolutionAns::eval_rpn(case_1.clone());
    SolutionAns::eval_rpn(case_2.clone());
}
